presto-memory-context/src/main/java/com/facebook/presto/memory/LocalMemoryContext.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import static com.facebook.presto.memory.MemoryReservationHandler.NOT_BLOCKED;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class LocalMemoryContext
+{
+    @Nonnull
+    private final AggregatedMemoryContext parentMemoryContext;
+    @GuardedBy("this")
+    private long usedBytes;
+
+    public LocalMemoryContext(AggregatedMemoryContext parentMemoryContext)
+    {
+        this.parentMemoryContext = requireNonNull(parentMemoryContext);
+    }
+
+    public synchronized long getBytes()
+    {
+        return usedBytes;
+    }
+
+    public synchronized ListenableFuture<?> setBytes(long bytes)
+    {
+        if (bytes == usedBytes) {
+            return NOT_BLOCKED;
+        }
+
+        // update the parent first as it may throw a runtime exception (e.g., ExceededMemoryLimitException)
+        ListenableFuture<?> future = parentMemoryContext.updateBytes(bytes - usedBytes);
+        usedBytes = bytes;
+        return future;
+    }
+
+    public synchronized ListenableFuture<?> addBytes(long bytes)
```

Comment:
I know we've talked about this before. I'm still against the idea of having `addBytes` (or `tryReserveMemory`) in local memory context. If you need a try method in `LocalMemoryContext`, I would find `trySetBytes` reasonable.

Anyways, let me finish a pass. We can come back to this by that point．